src/algebra/monoid_algebra.lean

```diff
@@ -250,6 +251,86 @@ finsupp.comap_distrib_mul_action_self
 
 end derived_instances
 
+/-! #### Non-unital, non-associative algebra structure -/
+section non_unital_non_assoc_algebra
+
+variables (k) [semiring k] [has_mul G]
+
+instance is_scalar_tower_self : is_scalar_tower k (monoid_algebra k G) (monoid_algebra k G) :=
+⟨λ t a b, show (t • a) * b = t • (a * b),
+begin
+  ext m,
+  classical,
+  simp only [mul_def, finsupp.mul_sum, smul_apply, sum_smul_index, implies_true_iff,
+    eq_self_iff_true, single_zero, finsupp.sum_apply, smul_eq_mul, sum_zero, zero_mul, mul_assoc,
+    single_apply, mul_ite, mul_zero],
+end⟩
+
+/-- If the coefficients are commutative amongst themselves, the also commute with the algebra
+multiplication. -/
+instance smul_comm_class_self {k : Type u₁} [comm_semiring k] :
+  smul_comm_class k (monoid_algebra k G) (monoid_algebra k G) :=
+⟨λ t a b, show t • (a * b) = a * (t • b),
+begin
+  ext m,
+  classical,
+  simp only [mul_def, finsupp.mul_sum, smul_apply, sum_smul_index, implies_true_iff,
```

Comment:
Are you sure you don't want to use `mul_apply` here, defined below? I think this stuff might still be too soon